tethys_layouts/views/map_layout.py

```diff
@@ -0,0 +1,850 @@
+"""
+********************************************************************************
+* Name: map_layout.py
+* Author: nswain
+* Created On: June 24, 2021
+* Copyright: (c) Aquaveo 2021
+********************************************************************************
+"""
+from abc import ABCMeta
+import collections
+from io import BytesIO
+import json
+import logging
+import os
+import requests
+import uuid
+from zipfile import ZipFile
+
+from django.core.exceptions import PermissionDenied
+from django.http import HttpResponse, JsonResponse
+from django.shortcuts import render
+import shapefile  # PyShp
+
+from tethys_layouts.exceptions import TethysLayoutPropertyException
+from tethys_layouts.mixins.map_layout import MapLayoutMixin
+from tethys_layouts.utilities import classproperty
+from tethys_layouts.views.tethys_layout import TethysLayout
+from tethys_sdk.permissions import has_permission
+from tethys_sdk.gizmos import ToggleSwitch, CesiumMapView, MapView, MVView, SlideSheet, SelectInput
+
+log = logging.getLogger(f'tethys.{__name__}')
+
+
+class MapLayout(TethysLayout, MapLayoutMixin):
```

Comment:
Other than segmenting code, is there a reason why the `MapLayoutMixin` needs to be separate from the `MapLayout` class? Do you anticipate the `MapLayoutMixin` being used elsewhere?

Reply:
The short answer is Yes. The methods contained in the MapLayoutMixin were originally in a separate object in ATCore called a MapManager. I made the decision to merge them into the MapLayout class to simplify the workflow a bit. In ATCore you would need to define a MapManager that would be responsible for composing the layers of your map and then the MapLayout would use it to compose the map. The advantage to this was that the MapManager could be used in other contexts (e.g. Condor workflows) without all the baggage of the MapLayout.

In the Tethys implementation, I wanted to simplify the process as much as possible. So now the methods for creating the layers are part of the MapLayout through this mixin and you define the layers directly in the MapLayout `compose_layers` method. Keeping the methods as a separate mixin will allow for the MapManager pattern to be implemented still, but now optional. ATCore will probably the mixin to maintain the MapManager pattern when we migrate after Tethys 4 is released.